apps/prairielearn/src/lib/group-update.js

```diff
@@ -1,336 +1,207 @@
-                  job.verbose(`----------------------------------------`);
-                  job.verbose(
-                    `ERROR: The following users do not exist. Please check their uids first.`,
-                  );
-                  notExist.forEach((user) => {
-                    job.verbose(user);
-                  });
-                  errorCount += notExist.length;
-                }
-                if (inGroup) {
-                  job.verbose(`----------------------------------------`);
-                  job.verbose(`ERROR: The following users are already in a group.`);
-                  inGroup.forEach((user) => {
-                    job.verbose(user);
-                  });
-                  errorCount += inGroup.length;
-                }
-                successCount = allCount - errorCount;
-              }
-              job.verbose(`----------------------------------------`);
-              if (errorCount === 0) {
-                job.verbose(
-                  `Successfully updated groups for ${successCount} students, with no errors`,
-                );
-                job.succeed();
-              } else {
-                job.verbose(`Successfully updated groups for ${successCount} students`);
-                job.fail(`Error updating ${errorCount} students`);
-              }
-            })()
-              .catch((err) => {
-                logger.error('Error while creating groups', err);
-                serverJobs.failJobSequence(job_sequence_id);
-              })
-              .finally(() => {
-                namedLocks.releaseLock(lock, (lockErr) => {
-                  if (lockErr) {
-                    job.fail(`ERROR: The lock ${lockName} was not released successfully`);
-                    return;
-                  }
-                  job.verbose(`Released lock ${lockName}`);
-                });
-              });
-          });
+    serverJob.executeInBackground(async (job) => {
+      const lockName = 'grouping assessment_id ' + assessment_id;
+      job.verbose(`Trying lock ${lockName}`);
+      await namedLocks.doWithLock(lockName, { timeout: 10000 }, async () => {
+        job.verbose(`Acquired lock ${lockName}`);
+        job.verbose('Auto generate group settings for ' + assessmentLabel);
+        job.verbose(`----------------------------------------`);
+        job.verbose(`Fetching the enrollment lists...`);
+        const resultList = await sqldb.queryAsync(sql.select_enrollments, { assessment_id });
+        const students = resultList.rows.map((row) => row.user_list);
+        _.shuffle(students);
+        const numStudents = resultList.rowCount;
+        const notAssigned = [];
+        const resultList2 = await sqldb.queryAsync(sql.select_not_assigned, { assessment_id });
+        resultList2.rows.forEach((element) => {
+          notAssigned.push(element.user_list);
         });
+        _.shuffle(notAssigned);
+        const numNotAssigned = resultList2.rowCount;
```

Comment:
This code does nothing useful, can be removed (unless this is a precursor to implement grouping students that may already have some groups in place).

Reply:
Are you talking about everything from `select_not_assigned` to printing `... of them have not been in a group`?

Reply:
Yes, in essence anything related to `notAssigned` and `numNotAssigned` is not in use. So the line before `select_not_assigned` (the declaration of `notAssigned`) too.

Reply:
I removed this code in c71bdd7.